hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/lock/LockManager.java

```diff
@@ -25,42 +25,146 @@
+   *
+   * @param resource on which the lock has to be acquired
+   * @deprecated Use {@link LockManager#writeLock} instead
+   */
+  public void lock(final R resource) {
+   writeLock(resource);
+  }
+
+  /**
+   * Releases the lock on given resource.
+   *
+   * @param resource for which the lock has to be released
+   * @deprecated Use {@link LockManager#writeUnlock} instead
+   */
+  public void unlock(final R resource) {
+   writeUnlock(resource);
+  }
+
+  /**
+   * Acquires the read lock on given resource.
+   *
+   * <p>Acquires the read lock on resource if the write lock is not held by
+   * another thread and returns immediately.
+   *
+   * <p>If the write lock on resource is held by another thread then
+   * the current thread becomes disabled for thread scheduling
+   * purposes and lies dormant until the read lock has been acquired.
+   *
+   * @param resource on which the read lock has to be acquired
+   */
+  public void readLock(final R resource) {
+    acquire(resource, ActiveLock::readLock);
+  }
+
+  /**
+   * Releases the read lock on given resource.
+   *
+   * @param resource for which the read lock has to be released
+   * @throws IllegalMonitorStateException if the current thread does not
+   *                                      hold this lock
+   */
+  public void readUnlock(final R resource) throws IllegalMonitorStateException {
+    release(resource, ActiveLock::readUnlock);
+  }
+
+  /**
+   * Acquires the write lock on given resource.
+   *
+   * <p>Acquires the write lock on resource if neither the read nor write lock
+   * are held by another thread and returns immediately.
+   *
+   * <p>If the current thread already holds the write lock then the
+   * hold count is incremented by one and the method returns
+   * immediately.
+   *
+   * <p>If the lock is held by another thread then the current
+   * thread becomes disabled for thread scheduling purposes and
+   * lies dormant until the write lock has been acquired.
+   *
+   * @param resource on which the lock has to be acquired
    */
-  public void lock(T resource) {
-    activeLocks.compute(resource, (k, v) -> {
-      ActiveLock lock;
+  public void writeLock(final R resource) {
+    acquire(resource, ActiveLock::writeLock);
+  }
+
+  /**
+   * Releases the write lock on given resource.
+   *
+   * @param resource for which the lock has to be released
+   * @throws IllegalMonitorStateException if the current thread does not
+   *                                      hold this lock
+   */
+  public void writeUnlock(final R resource) throws IllegalMonitorStateException {
+    release(resource, ActiveLock::writeUnlock);
+  }
+
+  /**
+   * Acquires the lock on given resource using the provided lock function.
+   *
+   * @param resource on which the lock has to be acquired
+   * @param lockFn function to acquire the lock
+   */
+  private void acquire(final R resource, final Consumer<ActiveLock> lockFn) {
+    lockFn.accept(getLockForLocking(resource));
```

Comment:
Is there any reason why acquire and release have different way of logic:
Release: Get lock for release. Release lock. Decrement count.
Acquire: In getlockfor acquire increment count, and then acquire lock. Why not follow similar for release. GetLockForAcquire, acquire lock, increment count. So, that it will be clear during code read.

Reply:
While acquiring the lock if we don't increment the active count as part of getLock call (atomically) we will end up in inconsistent state.

Let's say we got the lock and didn't increment the count, we try to acquire the lock and some other thread is holding the same lock. The current thread will wait, we still haven't incremented the active counter. 
While the other thread which was holding the lock will release the lock, as part of releasing we check the active lock count to remove lock object from the map and return the lock object to pool. In this case the active count will be 0 and we will end up removing the lock object from map, but the same lock object is given out.
Now we have a lock which is given out, but the lock object is removed from the map and returned to the pool.

Reply:
Understood. Thanks for the clear explanation. If it is not much work, can we add some comments in code it will be easy when reading? (If you think this is obvious, you can leave it).

Reply:
Added Additional comment for clarity.